incubator/hnc/internal/validators/hncconfig.go

```diff
@@ -141,6 +147,160 @@ func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationS
 	return allow("")
 }
 
+// objSet is a set of namespaced objects. It also stores if the object has been
+// checked for naming conflict.
+type objSet map[string]inNamespace
+type inNamespace struct {
+	namespace string
+	checked   bool
+}
+
+// conflicts stores object naming conflicts by GVK, object name and groups of
+// conflicting namespaces. The namespaces are grouped by the first ancestor.
+type conflicts map[schema.GroupVersionKind]map[string]affectedNamespaces
+type affectedNamespaces map[string][]string
+
+func (c *HNCConfig) checkForest(inst *api.HNCConfiguration) admission.Response {
+	c.Forest.Lock()
+	defer c.Forest.Unlock()
+
+	// Get types that are changed from other modes to "Propagate" mode.
+	gvks := c.getNewPropagateTypes(inst)
+
+	// Check if user-created objects would be overwritten by these mode changes.
+	confs := conflicts{}
+	for gvk, _ := range gvks {
+		// We will keep one object set for each type, because objects with the same
+		// name but different types won't affect each other.
+		obs := objSet{}
+
+		// Traverse all the namespaces.
+		for _, nnm := range c.Forest.GetNamespaceNames() {
+			ns := c.Forest.Get(nnm)
+
+			// Traverse all the original objects in each namespace. Please note that
```

Comment:
I found this section pretty hard to read. Also, I'm not sure that it will work with exceptions. How about doing something like this?

```
c := []conflicts{}
foreach ns {
  c = append(c, getConflictsFromTree(ns)
}

func getConflictsFromTree(ns) []conflicts {
  c := []conflicts{}
  foreach object in ns
    // when we get exceptions, hasConflictingAncestor will take the current
    // ns as a param too
    if src := hasConflictingAncestor(object); src != nil {
      c = append(c, src)
    }
  return c
}
```